libs/lb/lb-rs/src/text/buffer.rs

```diff
@@ -0,0 +1,583 @@
+        let ops = diff(&self.external.text, &text);
+
+        self.ops
+            .meta
+            .extend(ops.iter().map(|_| OpMeta { timestamp, base }));
+        self.ops.all.extend(ops.into_iter().map(Operation::Replace));
+
+        self.external.text = text;
+        self.external.seq = self.base.seq + self.ops.all.len();
+    }
+
+    /// Indicates to the buffer the changes that have been saved outside the editor. This will serve as the new base
+    /// for merging external changes. The sequence number should be taken from `current.seq` of the buffer when the
+    /// buffer's contents are read for saving.
+    pub fn saved(&mut self, external_seq: usize, external_text: String) {
+        self.external.text = external_text;
+        self.external.seq = external_seq;
+    }
+
+    pub fn merge(mut self, external_text_a: String, external_text_b: String) -> String {
+        let ops_a = diff(&self.external.text, &external_text_a);
+        let ops_b = diff(&self.external.text, &external_text_b);
+
+        let timestamp = Instant::now();
+        let base = self.external.seq;
+        self.ops
+            .meta
+            .extend(ops_a.iter().map(|_| OpMeta { timestamp, base }));
+        self.ops
+            .meta
+            .extend(ops_b.iter().map(|_| OpMeta { timestamp, base }));
+
+        self.ops
+            .all
+            .extend(ops_a.into_iter().map(Operation::Replace));
+        self.ops
+            .all
+            .extend(ops_b.into_iter().map(Operation::Replace));
+
+        self.update();
+        self.current.text
+    }
+
+    /// Applies all operations in the buffer's input queue
+    pub fn update(&mut self) -> Response {
+        // clear redo stack
+        //             v base        v current    v processed
+        // ops before: |<- applied ->|<- undone ->|<- queued ->|
+        // ops after:  |<- applied ->|<- queued ->|
+        let queue_len = self.base.seq + self.ops.len() - self.ops.processed_seq;
+        if queue_len > 0 {
+            let drain_range = self.current.seq..self.ops.processed_seq;
+            self.ops.all.drain(drain_range.clone());
+            self.ops.meta.drain(drain_range.clone());
+            self.ops.transformed.drain(drain_range.clone());
+            self.ops.transformed_inverted.drain(drain_range.clone());
+            self.ops.processed_seq = self.current.seq;
+        } else {
+            return Response::default();
+        }
+
+        // transform & apply
+        let mut result = Response::default();
+        for idx in self.current_idx()..self.current_idx() + queue_len {
+            let mut op = self.ops.all[idx].clone();
+            let meta = &self.ops.meta[idx];
+            self.transform(&mut op, meta);
+            self.ops.transformed_inverted.push(self.current.invert(&op));
+            self.ops.transformed.push(op.clone());
+            self.ops.processed_seq += 1;
+
+            result |= self.redo();
+        }
```

Comment:
subtle point: ops must be transformed by ops that are themselves already transformed. There are a few right ways and a few wrong ways to write this nested for-loop (outer loop in `update()` here, inner loop in `transform()` below)